### ISSUE TITLE
unreachable regression with generic Callable

### ISSUE THREAD

**Issue by dpinol** [python/mypy#21182]
**Bug Report**

Since mypy 1.20.0, a comparison between 2 generic callables may cause  "Statement is unreachable"

**To Reproduce**

```python
from typing import Callable

def identity[T](x: T) -> T:
    return x

def msg[T,S](cmp_property: Callable[[T], S]) -> None:
    if cmp_property == identity:
        return
```

**Expected Behavior**

Mypy should not report any error.

**Actual Behavior**

```
scpytools/kk.py: note: In function "msg":
scpytools/kk.py:10: error: Statement is unreachable  [unreachable]
            return
            ^~~~~~
```

**Your Environment**

- Mypy version used: mypy 1.20.0 (compiled: yes)
- Mypy command-line flags: mypy --warn-unreachable
- Mypy configuration options from `mypy.ini` (and other config files): No config files
- Python version used: 3.14.2

**Comment by hauntsaninja** [comment]
Thanks, changes in 1.20 surface a pre-existing deficiency in `is_overlapping_types`, will look more...

**Comment by hauntsaninja** [comment]
One idea for a fix is this:
```
--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -548,6 +551,14 @@ def is_overlapping_types(
             is_proper_subtype=False,
             ignore_pos_arg_names=not overlap_for_overloads,
             allow_partial_overlap=True,
+        ) or is_callable_compatible(
+            right,
+            left,
+            is_compat=_is_overlapping_types,
+            is_proper_subtype=False,
+            ignore_pos_arg_names=not overlap_for_overloads,
+            check_args_covariantly=True,
+            allow_partial_overlap=True,
         )
```

This is nice because it obviously solves any other symmetry issues, but maybe it is a bandaid over some deeper issue

**Comment by hauntsaninja** [comment]
Okay, it looks like this is known and we do something similar elsewhere, see e.g. https://github.com/python/mypy/pull/5476/changes#diff-f96a2d6138bc6cdf2a07c4d37f6071cc25c1631afc107e277a28d5b59fc0ef04R3755 so I will open the PR
